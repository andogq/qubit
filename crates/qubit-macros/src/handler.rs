use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote, ToTokens};
use syn::{
    meta::ParseNestedMeta, spanned::Spanned, Error, FnArg, ItemFn, Pat, Result, ReturnType, Type,
    TypeImplTrait,
};

/// Handlers can have different variations depending on how they interact with the client.
pub enum HandlerKind {
    /// Query handlers support the standard request/response pattern.
    Query,

    /// Subscriptions have an initial request, and returns a stream of responses that the client
    /// will continue to consume.
    Subscription,
}

impl HandlerKind {
    /// Attempt to parse the handler kind from [`ParseNestedMeta`].
    pub fn parse(&mut self, meta: ParseNestedMeta) -> Result<()> {
        if meta.path.is_ident("query") {
            *self = Self::Query;
            Ok(())
        } else if meta.path.is_ident("subscription") {
            *self = Self::Subscription;
            Ok(())
        } else {
            Err(meta.error("unsupported handler property"))
        }
    }
}

/// Generates the implementation for [`qubit::Handler`] for the provided handler function. The
/// [`HandlerKind`] is required alter how the handler is applied to the router. This could be
/// induced based on the return type of the handler (whether it retrusn a [`futures::Stream`]) or
/// not), but that might cause problems.
pub fn generate_handler(handler: ItemFn, kind: HandlerKind) -> Result<TokenStream> {
    let span = handler.span().clone();

    // Handlers must be async
    if handler.sig.asyncness.is_none() {
        return Err(Error::new_spanned(handler, "RPC handlers must be async"));
    }

    // Clone the function implementation, in order to use it as the handler
    let handler_fn = {
        let mut f = handler.clone();
        f.sig.ident = Ident::new("handler", Span::call_site());
        f
    };

    // Extract out the function name
    let function_name_str = handler.sig.ident.to_string();
    let function_ident = handler.sig.ident;

    // Extract out the return type
    let return_type = match handler.sig.output.clone() {
        ReturnType::Default => None,
        ReturnType::Type(_, ty) => Some(match *ty {
            Type::ImplTrait(TypeImplTrait { bounds, .. }) => {
                quote! { <dyn #bounds as futures::Stream>::Item }
            }
            ref return_type => return_type.to_token_stream(),
        }),
    };

    let mut inputs = handler.sig.inputs.iter();

    let ctx_ty = if let Some(FnArg::Typed(arg)) = inputs.next() {
        arg.ty.clone()
    } else {
        return Err(syn::Error::new(span, "ctx type must be provided"));
    };

    // Process parameters, to get the idents, string version of the idents, and the type
    let ((param_names, param_name_strs), param_tys): ((Vec<_>, Vec<_>), Vec<_>) = inputs
        .map(|param| match param {
            FnArg::Typed(arg) => match arg.pat.as_ref() {
                Pat::Ident(ident) => Ok(((ident.clone(), ident.ident.to_string()), arg.ty.clone())),
                Pat::Struct(_) | Pat::Tuple(_) | Pat::TupleStruct(_) => Err(Error::new(
                    arg.span(),
                    "destructured arguments are not currently supported",
                )),
                _ => Err(Error::new(
                    arg.span(),
                    "unable to process this argument type",
                )),
            },
            FnArg::Receiver(_) => Err(Error::new(
                param.span(),
                "handlers cannot have `self` as a parameter",
            )),
        })
        .collect::<Result<Vec<_>>>()?
        .into_iter()
        .unzip();

    let parse_params = (!param_names.is_empty()).then(|| {
        quote! {
            let (#(#param_names,)*) = params.parse::<(#(#param_tys,)*)>().unwrap();
        }
    });

    let (register_impl, signature) = match kind {
        HandlerKind::Query => (
            quote! {
                rpc_builder.query(#function_name_str, |app_ctx, params| async move {
                    #parse_params

                    // Convert app_ctx to ctx
                    let ctx = <#ctx_ty as qubit::FromContext<__internal_AppCtx>>::from_app_ctx(app_ctx).unwrap();

                    // Run the handler
                    let result = handler(ctx, #(#param_names,)*).await;

                    // Serialise the resulte
                    serde_json::to_value(result).unwrap()
                })
            },
            {
                let return_type = match return_type.as_ref() {
                    Some(return_type) => {
                        quote!(<#return_type as ts_rs::TS>::name())
                    }
                    None => quote!("void"),
                };

                quote! {
                    format!("({}) => Promise<{}>", parameters.join(", "), #return_type)
                }
            },
        ),
        HandlerKind::Subscription => {
            let notification_name = format!("{function_name_str}_notif");
            let unsubscribe_name = format!("{function_name_str}_unsub");

            (
                quote! {
                    rpc_builder.subscription(#function_name_str, #notification_name, #unsubscribe_name, |app_ctx, params| async move {
                        #parse_params

                        // Convert app_ctx to ctx
                        let ctx = <#ctx_ty as qubit::FromContext<__internal_AppCtx>>::from_app_ctx(app_ctx).unwrap();

                        // Run the handler
                        let stream = handler(ctx, #(#param_names,)*).await;

                        futures::StreamExt::map(
                            stream,
                            |value| serde_json::to_value(value).unwrap()
                        )
                    })
                },
                {
                    let Some(return_type) = return_type.as_ref() else {
                        return Err(syn::Error::new(
                            span,
                            "subscriptions must have a return type",
                        ));
                    };

                    quote! {
                        format!("({}) => Stream<{}>", parameters.join(", "), <#return_type as ts_rs::TS>::name())
                    }
                },
            )
        }
    };

    Ok(quote! {
        #[allow(non_camel_case_types)]
        struct #function_ident;
        impl<__internal_AppCtx> qubit::Handler<__internal_AppCtx> for #function_ident
            where #ctx_ty: qubit::FromContext<__internal_AppCtx>,
                __internal_AppCtx: 'static + Send + Sync + Clone
        {
            fn get_type() -> qubit::HandlerType {
                let parameters = [
                    #((#param_name_strs, <#param_tys as ts_rs::TS>::name())),*
                ]
                    .into_iter()
                    .map(|(param, ty): (&str, String)| {
                        format!("{param}: {ty}")
                    })
                    .collect::<Vec<_>>();

                qubit::HandlerType {
                    name: #function_name_str.to_string(),
                    signature: #signature,
                }
            }

            fn register(rpc_builder: qubit::RpcBuilder<__internal_AppCtx>) -> qubit::RpcBuilder<__internal_AppCtx> {
                #handler_fn

                #register_impl
            }

            fn add_dependencies(dependencies: &mut std::collections::BTreeMap<std::string::String, std::string::String>) {
                // Add dependencies for the parameters
                #(<#param_tys as qubit::TypeDependencies>::get_deps(dependencies);)*

                // Add dependencies for the return type
                <#return_type as qubit::TypeDependencies>::get_deps(dependencies);
            }
        }
    })
}
