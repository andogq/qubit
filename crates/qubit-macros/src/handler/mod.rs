use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote, ToTokens};
use syn::{
    parse_quote, spanned::Spanned, Error, FnArg, ItemFn, Pat, Result, ReturnType, Type,
    TypeImplTrait, Visibility,
};

mod options;

pub use options::*;

/// Kind of return value from a handler.
enum HandlerReturn {
    /// Handler returns a stream of the provided type.
    Stream(Type),

    /// Handle returns a single instance of the provided type.
    Return(Type),
}

impl HandlerReturn {
    /// Get the equivalent TS type for this return type.
    fn ts_container(&self) -> String {
        match self {
            Self::Stream(_) => "Stream".to_string(),
            Self::Return(_) => "Promise".to_string(),
        }
    }

    /// Register any required inbuilt types.
    fn register_inbuilt(&self, registry: &Ident) -> TokenStream {
        match self {
            HandlerReturn::Stream(_) => {
                quote! { #registry.inbuilt(qubit::builder::ty::InbuiltType::Stream); }
            }
            HandlerReturn::Return(_) => TokenStream::new(),
        }
    }
}

impl ToTokens for HandlerReturn {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            Self::Stream(ty) | Self::Return(ty) => ty.to_tokens(tokens),
        }
    }
}

/// All relevant information about the handler extracted from the macro.
pub struct Handler {
    /// Visibility of the handler.
    visibility: Visibility,

    /// Name of the handler.
    name: Ident,

    /// Type of the context used in the handler.
    ctx_ty: Type,

    /// Inputs for the handler. Currently does not support any kind of destructuring.
    inputs: Vec<(Ident, Type)>,

    /// Return type of the handler.
    return_type: HandlerReturn,

    /// The actual handler implementation.
    implementation: ItemFn,
}

impl Handler {
    /// Parse a handler from an [`ItemFn`] and some options. This will return [`syn::Error`]s if
    /// parsing cannot take place.
    pub fn parse(handler: ItemFn, options: HandlerOptions) -> Result<Self> {
        let span = handler.span();

        // TODO: Could this be relaxed?
        if handler.sig.asyncness.is_none() {
            return Err(Error::new(span, "handlers must be async"));
        }

        let implementation = {
            // Create the implementation by cloning the original function, and changing the
            // name to be `handler`.
            let mut implementation = handler.clone();
            implementation.sig.ident = Ident::new("handler", Span::call_site());
            implementation
        };

        let mut inputs = handler
            .sig
            .inputs
            .into_iter()
            .map(|arg| {
                let FnArg::Typed(arg) = arg else {
                    return Err(Error::new(span, "handlers cannot take `self` parameter"));
                };

                let Pat::Ident(ident) = *arg.pat else {
                    return Err(Error::new(
                        span,
                        "destructured parameters are not supported in handlers",
                    ));
                };

                let ty = *arg.ty;

                Ok((ident.ident, ty))
            })
            .collect::<Result<Vec<_>>>()?;

        // TODO: Remove this restriction, allow handlers to accept no parameters
        if inputs.is_empty() {
            return Err(Error::new(
                span,
                "handlers must accept atleast one argument (the ctx)",
            ));
        }

        let (_, ctx_ty) = inputs.remove(0);

        Ok(Self {
            implementation,
            visibility: handler.vis,
            name: options.name.unwrap_or(handler.sig.ident),
            ctx_ty,
            inputs,
            return_type: {
                let return_type = match handler.sig.output {
                    ReturnType::Default => HandlerReturn::Return(parse_quote! { () }),
                    ReturnType::Type(_, ty) => match *ty {
                        // BUG: Assuming that any trait implementation is a stream, which definitely isn't
                        // the case.
                        Type::ImplTrait(TypeImplTrait { bounds, .. }) => HandlerReturn::Stream(
                            parse_quote! { <dyn #bounds as futures::Stream>::Item },
                        ),
                        // All other return types will be treated as a regular return type.
                        return_type => HandlerReturn::Return(return_type),
                    },
                };

                match (&return_type, options.kind) {
                    // Valid case, return type matches with handler annotation
                    (HandlerReturn::Stream(_), Some(HandlerKind::Subscription))
                    | (HandlerReturn::Return(_), Some(HandlerKind::Query) | None) => return_type,

                    // Mismatches
                    (HandlerReturn::Stream(_), Some(HandlerKind::Query) | None) => {
                        return Err(Error::new(
                            span,
                            "handler indicated to be a query, however a stream was returned",
                        ));
                    }
                    (HandlerReturn::Return(_), Some(HandlerKind::Subscription)) => {
                        return Err(Error::new(
                            span,
                            "handler indicated to be a subscription, however a stream was not returned",
                        ));
                    }
                }
            },
        })
    }

    /// Produce a list of parameter names as idents that this handler accepts.
    fn parameter_names(&self) -> Vec<Ident> {
        self.inputs.iter().map(|(name, _)| name).cloned().collect()
    }

    /// Produce a list of parameter names as strings that this handler accepts.
    fn parameter_names_str(&self) -> Vec<String> {
        self.parameter_names()
            .iter()
            .map(|name| name.to_string())
            .collect()
    }

    /// Produce a list of parameter types that this handler accepts.
    fn parameter_tys(&self) -> Vec<Type> {
        self.inputs.iter().map(|(_, ty)| ty).cloned().collect()
    }

    /// Produce a token stream that will generate the TS signature of this handler.
    fn get_signature(&self) -> TokenStream {
        let return_type = &self.return_type;
        let container = return_type.ts_container();

        let param_names_str = self.parameter_names_str();
        let param_tys = self.parameter_tys();

        quote! {
            {
                let parameters = [
                    #((#param_names_str, <#param_tys as ts_rs::TS>::name())),*
                ]
                    .into_iter()
                    .map(|(param, ty): (&str, String)| {
                        format!("{param}: {ty}")
                    })
                    .collect::<Vec<_>>()
                    .join(", ");

                format!("({}) => {}<{}>", parameters, #container, <#return_type as ts_rs::TS>::name())
            }
        }
    }
}

impl From<Handler> for TokenStream {
    fn from(handler: Handler) -> Self {
        // Generate the signature
        let param_names = handler.parameter_names();
        let param_tys = handler.parameter_tys();
        let signature = handler.get_signature();

        // Extract required elements from handler
        let Handler {
            visibility,
            name,
            ctx_ty,
            inputs,
            return_type,
            implementation,
        } = handler;

        let handler_name_str = name.to_string();

        let register_impl = {
            // Define idents in one place, so they will be checked by the compiler
            let ctx_ident = quote! { ctx };
            let params_ident = quote! { params };

            // Generate the parameter parsing implementation
            let parse_params = (!inputs.is_empty()).then(|| {
                quote! {
                    let (#(#param_names,)*) = #params_ident.parse::<(#(#param_tys,)*)>().unwrap();
                }
            });

            // Body of the handler registration implementation
            let register_inner = quote! {
                #parse_params

                handler(#ctx_ident, #(#param_names,)*).await
            };

            match &return_type {
                HandlerReturn::Return(_) => {
                    quote! {
                        rpc_builder.query(#handler_name_str, |#ctx_ident, #params_ident| async move {
                            #register_inner
                        })
                    }
                }
                HandlerReturn::Stream(_) => {
                    let notification_name = format!("{handler_name_str}_notif");
                    let unsubscribe_name = format!("{handler_name_str}_unsub");

                    quote! {
                        rpc_builder.subscription(
                            #handler_name_str,
                            #notification_name,
                            #unsubscribe_name,
                            |#ctx_ident, #params_ident| async move {
                                #register_inner
                            }
                        )
                    }
                }
            }
        };

        // Must be a collision-free ident to use as a generic within the handler
        let inner_ctx_ty = quote! { __internal_AppCtx };

        let registry_ident = parse_quote! { registry };
        let inbuilt_return_type = return_type.register_inbuilt(&registry_ident);

        quote! {
            #[allow(non_camel_case_types)]
            #visibility struct #name;
            impl<#inner_ctx_ty> qubit::Handler<#inner_ctx_ty> for #name
                where #ctx_ty: qubit::FromContext<#inner_ctx_ty>,
                    #inner_ctx_ty: 'static + Send + Sync + Clone
            {
                fn get_type() -> qubit::HandlerType {
                    qubit::HandlerType {
                        name: #handler_name_str.to_string(),
                        signature: #signature,
                    }
                }

                fn register(rpc_builder: qubit::RpcBuilder<#inner_ctx_ty>) -> qubit::RpcBuilder<#inner_ctx_ty> {
                    #implementation

                    #register_impl
                }

                fn export_types(#registry_ident: &mut qubit::builder::ty::TypeRegistry) {
                    #inbuilt_return_type

                    // Add dependencies for the parameters
                    #(<#param_tys as qubit::ExportType>::export(#registry_ident);)*

                    // Add dependencies for the return type
                    <#return_type as qubit::ExportType>::export(#registry_ident);
                }
            }
        }.into()
    }
}
