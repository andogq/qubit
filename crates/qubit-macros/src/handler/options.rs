use std::fmt::Display;

use syn::{meta::ParseNestedMeta, Ident, LitStr, Result};

/// Handlers can have different variations depending on how they interact with the client.
#[derive(Clone)]
pub enum HandlerKind {
    /// Query handlers support the standard request/response pattern, and are safe to be cached.
    Query,

    /// Mutation handlers also support the standard request/response pattern, however they should
    /// not be cached.
    Mutation,

    /// Subscriptions have an initial request, and returns a stream of responses that the client
    /// will continue to consume.
    Subscription,
}

impl Display for HandlerKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                HandlerKind::Query => "Query",
                HandlerKind::Mutation => "Mutation",
                HandlerKind::Subscription => "Subscription",
            }
        )
    }
}

/// Options that may be attached to a handler.
pub struct HandlerOptions {
    /// The kind of handler.
    pub kind: HandlerKind,

    /// Overridden name for the handler.
    pub name: Option<Ident>,
}

impl HandlerOptions {
    /// Build up an instance of [`HandlerOptions`].
    pub fn builder() -> HandlerOptionsBuilder {
        HandlerOptionsBuilder::default()
    }
}

/// Builder for [`HandlerOptions`]. Allows for the kind to be empty until it's provided.
#[derive(Default)]
pub struct HandlerOptionsBuilder {
    /// Kind of the handler.
    pub kind: Option<HandlerKind>,

    /// Overridden name of the handler.
    pub name: Option<Ident>,
}

impl HandlerOptionsBuilder {
    /// Attempt to parse the handler kind from [`ParseNestedMeta`].
    pub fn parse(&mut self, meta: ParseNestedMeta) -> Result<()> {
        if meta.path.is_ident("query") {
            self.kind = Some(HandlerKind::Query);
            Ok(())
        } else if meta.path.is_ident("mutation") {
            self.kind = Some(HandlerKind::Mutation);
            Ok(())
        } else if meta.path.is_ident("subscription") {
            self.kind = Some(HandlerKind::Subscription);
            Ok(())
        } else if meta.path.is_ident("name") {
            // Extract name from the attribute
            let name = meta.value()?.parse::<LitStr>()?.value();

            // Create the ident for the handler name
            let ident = Ident::new(&name, meta.input.span());

            self.name = Some(ident);
            Ok(())
        } else {
            Err(meta.error("unsupported handler property"))
        }
    }

    /// Consume the builder to produce [`HandlerOptions`]. Will return `None` if the builder was
    /// in an invalid state.
    pub fn build(self) -> Option<HandlerOptions> {
        Some(HandlerOptions {
            kind: self.kind?,
            name: self.name,
        })
    }
}
