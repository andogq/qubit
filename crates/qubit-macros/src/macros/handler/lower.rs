use quote::quote;
use syn::{Expr, Ident, ItemFn, parse_quote};

use super::{analyse::Model, parse::HandlerKind};

pub fn lower(model: Model) -> Ir {
    Ir {
        name: model.name,
        kind: {
            let variant = match model.kind {
                HandlerKind::Query => quote!(Query),
                HandlerKind::Mutation => quote!(Mutation),
                HandlerKind::Subscription => quote!(Subscription),
            };

            parse_quote!(::qubit::__private::HandlerKind::#variant)
        },
        rpc_name: model.rpc_name,
        param_names: model
            .param_names
            .into_iter()
            .map(|param| param.to_string())
            .collect(),
        handler: model.handler,
    }
}

pub struct Ir {
    pub name: Ident,
    pub kind: Expr,
    pub rpc_name: String,
    pub param_names: Vec<String>,
    pub handler: ItemFn,
}

#[cfg(test)]
mod test {
    use rstest::rstest;

    use super::super::analyse::ModelAssertion;

    use super::*;

    #[derive(Clone)]
    struct IrAssertion {
        name: Ident,
        kind: Expr,
        rpc_name: String,
        param_names: Vec<String>,
    }

    impl IrAssertion {
        fn new(name: Ident, kind: Expr) -> Self {
            Self {
                rpc_name: name.to_string(),
                name,
                kind,
                param_names: Vec::new(),
            }
        }

        fn query(name: Ident) -> Self {
            Self::new(name, parse_quote!(::qubit::__private::HandlerKind::Query))
        }

        fn mutation(name: Ident) -> Self {
            Self::new(
                name,
                parse_quote!(::qubit::__private::HandlerKind::Mutation),
            )
        }

        fn subscription(name: Ident) -> Self {
            Self::new(
                name,
                parse_quote!(::qubit::__private::HandlerKind::Subscription),
            )
        }

        fn with_rpc_name(mut self, rpc_name: impl ToString) -> Self {
            self.rpc_name = rpc_name.to_string();
            self
        }

        fn with_param_names(
            mut self,
            param_names: impl IntoIterator<Item = impl ToString>,
        ) -> Self {
            self.param_names = param_names
                .into_iter()
                .map(|param_name| param_name.to_string())
                .collect();
            self
        }
    }

    #[rstest]
    #[case::simple_query(
        ModelAssertion::query(parse_quote!(my_handler)),
        IrAssertion::query(parse_quote!(my_handler)),
    )]
    #[case::simple_mutation(
        ModelAssertion::mutation(parse_quote!(my_handler)),
        IrAssertion::mutation(parse_quote!(my_handler)),
    )]
    #[case::simple_subscription(
        ModelAssertion::subscription(parse_quote!(my_handler)),
        IrAssertion::subscription(parse_quote!(my_handler)),
    )]
    #[case::single_param(
        ModelAssertion::query(parse_quote!(my_handler))
            .with_param_names([parse_quote!(param_a)]),
        IrAssertion::query(parse_quote!(my_handler))
            .with_param_names(["param_a"]),
    )]
    #[case::multi_param(
        ModelAssertion::query(parse_quote!(my_handler))
            .with_param_names([parse_quote!(param_a), parse_quote!(param_b)]),
        IrAssertion::query(parse_quote!(my_handler))
            .with_param_names(["param_a", "param_b"]),
    )]
    #[case::with_rename(
        ModelAssertion::query(parse_quote!(my_handler))
            .with_rpc_name("other_name"),
        IrAssertion::query(parse_quote!(my_handler))
            .with_rpc_name("other_name"),
    )]
    fn valid(#[case] model: ModelAssertion, #[case] expected: IrAssertion) {
        let name = model.name;
        let ir = lower(Model {
            rpc_name: model.rpc_name,
            kind: model.kind,
            param_names: model.param_names,
            handler: parse_quote!(fn #name() {}),
            name,
        });

        assert_eq!(ir.name, expected.name);
        assert_eq!(ir.kind, expected.kind);
        assert_eq!(ir.rpc_name, expected.rpc_name);
        assert_eq!(ir.param_names, expected.param_names);
    }
}
