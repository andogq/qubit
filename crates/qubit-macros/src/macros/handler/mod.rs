use proc_macro2::TokenStream;
use syn::Error;

mod analyse;
mod codegen;
mod lower;
mod parse;

use self::{analyse::analyse, codegen::codegen, lower::lower, parse::parse};

pub fn handler(attrs: TokenStream, item: TokenStream) -> Result<TokenStream, Error> {
    let ast = parse(attrs, item)?;
    let model = analyse(ast)?;
    let ir = lower(model);
    Ok(codegen(ir))
}
