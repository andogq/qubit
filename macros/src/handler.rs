use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;
use syn::{
    meta::ParseNestedMeta, parse_quote, spanned::Spanned, Error, FnArg, ItemFn, Pat, Result,
    ReturnType,
};

/// Handlers can have different variations depending on how they interact with the client.
pub enum HandlerKind {
    /// Query handlers support the standard request/response pattern.
    Query,

    /// Subscriptions have an initial request, and returns a stream of responses that the client
    /// will continue to consume.
    Subscription,
}

impl HandlerKind {
    /// Attempt to parse the handler kind from [`ParseNestedMeta`].
    pub fn parse(&mut self, meta: ParseNestedMeta) -> Result<()> {
        if meta.path.is_ident("query") {
            *self = Self::Query;
            Ok(())
        } else if meta.path.is_ident("subscription") {
            *self = Self::Subscription;
            Ok(())
        } else {
            Err(meta.error("unsupported handler property"))
        }
    }
}

/// Generates the implementation for [`rs_ts_api::Handler`] for the provided handler function. The
/// [`HandlerKind`] is required alter how the handler is applied to the router. This could be
/// induced based on the return type of the handler (whether it retrusn a [`futures::Stream`]) or
/// not), but that might cause problems.
pub fn generate_handler(handler: ItemFn, kind: HandlerKind) -> Result<TokenStream> {
    let span = handler.span().clone();

    // Handlers must be async
    if handler.sig.asyncness.is_none() {
        return Err(Error::new_spanned(handler, "RPC handlers must be async"));
    }

    // Clone the function implementation, in order to use it as the handler
    let handler_fn = {
        let mut f = handler.clone();
        f.sig.ident = Ident::new("handler", Span::call_site());
        f
    };

    // Extract out the function name
    let function_name_str = handler.sig.ident.to_string();
    let function_ident = handler.sig.ident;

    // Extract out the return type
    let return_type = match handler.sig.output.clone() {
        ReturnType::Default => quote!("void"),
        ReturnType::Type(_, ty) => {
            quote!(<#ty as ts_rs::TS>::name())
        }
    };

    let mut inputs = handler.sig.inputs.iter();

    let ctx_ty = if let Some(FnArg::Typed(arg)) = inputs.next() {
        arg.ty.clone()
    } else {
        return Err(syn::Error::new(span, "ctx type must be provided"));
    };

    // Process parameters, to get the idents, string version of the idents, and the type
    let ((param_names, param_name_strs), param_tys): ((Vec<_>, Vec<_>), Vec<_>) = inputs
        .map(|param| match param {
            FnArg::Typed(arg) => match arg.pat.as_ref() {
                Pat::Ident(ident) => Ok(((ident.clone(), ident.ident.to_string()), arg.ty.clone())),
                Pat::Struct(_) | Pat::Tuple(_) | Pat::TupleStruct(_) => Err(Error::new(
                    arg.span(),
                    "destructured arguments are not currently supported",
                )),
                _ => Err(Error::new(
                    arg.span(),
                    "unable to process this argument type",
                )),
            },
            FnArg::Receiver(_) => Err(Error::new(
                param.span(),
                "handlers cannot have `self` as a parameter",
            )),
        })
        .collect::<Result<Vec<_>>>()?
        .into_iter()
        .unzip();

    let parse_params = (!param_names.is_empty()).then(|| {
        quote! {
            let (#(#param_names,)*) = params.parse::<(#(#param_tys,)*)>().unwrap();
        }
    });

    let register_impl = match kind {
        HandlerKind::Query => quote! {
            rpc_builder.query(#function_name_str, |app_ctx, params| async move {
                #parse_params

                // Convert app_ctx to ctx
                let ctx = <#ctx_ty as rs_ts_api::Context<__internal_AppCtx>>::from_app_ctx(app_ctx).unwrap();

                // Run the handler
                let result = handler(ctx, #(#param_names,)*).await;

                // Serialise the resulte
                serde_json::to_value(result).unwrap()
            })
        },
        HandlerKind::Subscription => {
            let notification_name = format!("{function_name_str}_notif");
            let unsubscribe_name = format!("{function_name_str}_unsub");

            quote! {
                rpc_builder.subscription(#function_name_str, #notification_name, #unsubscribe_name, |app_ctx, params| async move {
                    #parse_params

                    // Convert app_ctx to ctx

                    // Run the handler
                    handler(ctx, #(#param_names,)*)
                })
            }
        }
    };

    let return_ty = match handler.sig.output.clone() {
        ReturnType::Default => parse_quote!(()),
        ReturnType::Type(_, ty) => ty,
    };

    Ok(quote! {
        #[allow(non_camel_case_types)]
        struct #function_ident;
        impl<__internal_AppCtx> rs_ts_api::Handler<__internal_AppCtx> for #function_ident
            where #ctx_ty: rs_ts_api::Context<__internal_AppCtx>,
                __internal_AppCtx: 'static + Send + Sync + Clone
        {
            fn get_type() -> rs_ts_api::HandlerType {
                let parameters = [
                    #((#param_name_strs, <#param_tys as ts_rs::TS>::name())),*
                ]
                    .into_iter()
                    .map(|(param, ty): (&str, String)| {
                        format!("{param}: {ty}")
                    })
                    .collect::<Vec<_>>();

                rs_ts_api::HandlerType {
                    name: #function_name_str.to_string(),
                    signature: format!("({}) => Promise<{}>", parameters.join(", "), #return_type),
                }
            }

            fn register(rpc_builder: rs_ts_api::RpcBuilder<__internal_AppCtx>) -> rs_ts_api::RpcBuilder<__internal_AppCtx> {
                #handler_fn

                #register_impl
            }

            fn add_dependencies(dependencies: &mut std::collections::BTreeMap<std::string::String, std::string::String>) {
                // Add dependencies for the parameters
                #(<#param_tys as rs_ts_api::TypeDependencies>::get_deps(dependencies);)*

                // Add dependencies for the return type
                <#return_ty as rs_ts_api::TypeDependencies>::get_deps(dependencies);
            }
        }
    })
}
