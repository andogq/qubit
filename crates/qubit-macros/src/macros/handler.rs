use proc_macro2::Span;
use syn::{meta, parse_macro_input, spanned::Spanned, Error, Item};

use crate::handler::{Handler, HandlerOptions};

pub fn handler(
    attr: proc_macro::TokenStream,
    input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    // Extract information from the attribute
    let options = {
        let mut options_builder = HandlerOptions::builder();

        let attribute_parser = meta::parser(|meta| options_builder.parse(meta));

        parse_macro_input!(attr with attribute_parser);

        let Some(options) = options_builder.build() else {
            // Produce a compiler error
            // TODO: Make it a better erro
            return Error::new(
                Span::call_site(),
                "handler type must be provided (`query`, `mutation`, or `subscription`)",
            )
            .into_compile_error()
            .into();
        };

        options
    };

    // Attempt to match as a function
    syn::parse::<Item>(input)
        .and_then(|item| {
            if let Item::Fn(handler) = item {
                let handler = Handler::parse(handler, options)?;
                Ok(handler.into())
            } else {
                Err(Error::new(item.span(), "handlers must be a method"))
            }
        })
        .unwrap_or_else(Error::into_compile_error)
        .into()
}
