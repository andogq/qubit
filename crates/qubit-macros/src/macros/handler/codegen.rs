use proc_macro2::TokenStream;
use quote::quote;

use super::lower::Ir;

pub fn codegen(ir: Ir) -> TokenStream {
    let Ir {
        name,
        kind,
        rpc_name,
        param_names,
        handler,
    } = ir;

    quote! {
        #handler

        const _: () = {
            #[::qubit::__private::linkme::distributed_slice(::qubit::__private::HANDLER_DEFINITIONS)]
            #[linkme(crate = ::qubit::__private::linkme)]
            static HANDLER_DEFINITION: fn() -> (::core::any::TypeId, ::qubit::__private::HandlerMeta) = || (
                ::core::any::Any::type_id(&#name),
                ::qubit::__private::HandlerMeta {
                    kind: #kind,
                    name: #rpc_name,
                    param_names: &[#(#param_names),*]
                }
            );
        };
    }
}
