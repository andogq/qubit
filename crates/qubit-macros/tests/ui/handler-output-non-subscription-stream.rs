use qubit_macros::handler;

#[handler(query)]
async fn my_handler() -> impl Stream<Item = u32> {}

fn main() {}
