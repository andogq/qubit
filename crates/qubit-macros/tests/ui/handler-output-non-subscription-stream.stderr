error: handler isn't a subscription, but a stream was returned
 --> tests/ui/handler-output-non-subscription-stream.rs:4:26
  |
4 | async fn my_handler() -> impl Stream<Item = u32> {}
  |                          ^^^^^^^^^^^^^^^^^^^^^^^
